Add HAOC feature for deepin linux-6.6.y: Isolated Execution Environment(IEE)

amjac27 opened the pull request:
1. Add support for X86_64/AArch64 Isolated Execution Environment(IEE).
2. Add switch support in kernel command line: set "haoc=on" to enable HAOC.

## Summary by Sourcery

This pull request introduces the Isolated Execution Environment (IEE) feature for the deepin linux-6.6.y kernel, supporting both X86_64 and AArch64 architectures. The IEE provides an isolated environment for executing specific code, enhancing system security. It includes mechanisms for memory protection and access control.

New Features:
- Adds support for Isolated Execution Environment (IEE) on X86_64 and AArch64 architectures.
- Introduces a kernel command line switch "haoc=on" to enable the HAOC (Hardware-Assisted Out-of-Context) feature, which includes IEE.

Commit:
HAOC: Add support for x86 Isolated Execution Environment

community inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/IBQKOW

-------------------------------------------------

Support Isolated Execution Environment for x86.
IEE depends on CR0.wp

Signed-off-by: Shu Hang <shuh2023@zgclab.edu.cn>
Signed-off-by: Hu Bing <<hubing2023@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

Commit:
HAOC: Add support for AArch64 Isolated Execution Environment(IEE).

community inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/IBQKOW

-----------------------------------

The base framework of HAOC. Could isolate kernel critical data and
enforce all write access made and verified in IEE APIs.
Needs hardware support FEAT_HPDS.

Signed-off-by: Lyu Jinglin <lvjl2022@zgclab.edu.cn>
Signed-off-by: Zhang Shiyang <zhangsy2023@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

Commit:
HAOC: Add kernel command line support for x86 IEE.

community inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/IBQKOW

-----------------------------------

Use kernel command line haoc to control if HAOC should be enabled.
eg: haoc=on to enable haoc protection.

Signed-off-by: Shu Hang <shuh2023@zgclab.edu.cn>
Signed-off-by: Hu Bing <hubing2023@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

Commit:
HAOC: Add ARM64 kernel command line support to switch IEE on\off.

community inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/IBQKOW

-----------------------------------

Use kernel command line parameter "haoc" to control if HAOC should
be enabled on ARM64.

Signed-off-by: Lyu Jinglin <lvjl2022@zgclab.edu.cn>
Signed-off-by: Zhang Shiyang <zhangsy2023@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

amjac27 committed:
HAOC: Fix some build error on ARM64 in deepin linux-6.6.y

- change "__ptep_get" to "ptep_get"
- change "__set_pte" to "set_pte"
- change "P4D_TABLE_AF" to "PMD_SECT_AF"
- change the place of "iee_tcr_mask"

Signed-off-by: Lyu Jinglin <lvjl2022@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

amjac27 committed:
HAOC: fix memory recognization issue

community inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/IBTMTU

--------------------------------

add support for large memory (more than 1TB) in HAOC-ARM64 IEE.

Fixes: 8d2e6d1bb690 ("HAOC: Add support for AArch64 Isolated Execution Environment(IEE).")
Signed-off-by: Zhang Shiyang <zhangsy2023@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

Avenger-285714 commented on Makefile:
```diff
@@ -563,6 +563,7 @@ KBUILD_CFLAGS += -funsigned-char
 KBUILD_CFLAGS += -fno-common
 KBUILD_CFLAGS += -fno-PIE
 KBUILD_CFLAGS += -fno-strict-aliasing
+KBUILD_CFLAGS += -Wno-unused-variable
```
拒绝。不要有未使用的变量出现，实在必须要有，请将该CFLAG限定在自己的代码段之内，不要污染对其它代码的检查。

Avenger-285714 commented on arch/arm64/configs/deepin_arm64_desktop_defconfig:
```diff
@@ -4510,3 +4510,8 @@ CONFIG_RCU_CPU_STALL_TIMEOUT=60
+# Hardware Assisted OS Compartmentalization(HAOC)
+#
+CONFIG_IEE=y
+# end of Hardware Assisted OS Compartmentalization(HAOC)
```
请使用"make savedefconfig"命令修改defconfig而非手动更改。

Avenger-285714 commented on arch/arm64/kernel/Makefile:
```diff
@@ -36,6 +36,7 @@ obj-y			:= debug-monitors.o entry.o irq.o fpsimd.o		\
 			   syscall.o proton-pack.o idreg-override.o idle.o	\
 			   patching.o
 
+obj-$(CONFIG_IEE)				+= haoc/
```
请对齐缩进。

Avenger-285714 commented on arch/arm64/kernel/haoc/Makefile:
```diff
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: GPL-2.0
+obj-y += haoc.o
+obj-y += iee/
```
末尾添加空行。

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/Makefile:
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: GPL-2.0
+obj-$(CONFIG_IEE) += iee.o iee-gate.o iee-init.o iee-func.o iee-mmu.o
```
末尾添加空行。

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/iee.c:
```diff
@@ -0,0 +1,32 @@
+ * with outer API like iee_memset();
+ */
+#pragma GCC push_options
+#pragma GCC optimize("O0")
```
此处为何使用O0

Jinglin-lyu replied:
这个是一个 IEE 函数示例。我们想写一个 IEE 内部使用的 memset 函数代码，但是编译器在识别到类似 memset 操作时会自动优化为对 memset 库函数的调用，而 IEE 调用外部函数访问数据会引发安全性问题，因此必须防止该函数被执行优化。

Avenger-285714 replied:
> 这个是一个 IEE 函数示例。我们想写一个 IEE 内部使用的 memset 函数代码，但是编译器在识别到类似 memset 操作时会自动优化为对 memset 库函数的调用，而 IEE 调用外部函数访问数据会引发安全性问题，因此必须防止该函数被执行优化。

试一下用volatile声明该函数？"#pragma GCC push_options"这样的用法会导致clang编译器编不过。

Jinglin-lyu replied:
无法使用 volatile 声明函数，但可以改一下写法让它适配 clang。
#if defined(__clang__)
__attribute__((optnone))
#elif defined(__GNUC__) || defined(__GNUG__)
__attribute__((optimize("O0")))
#endif

Avenger-285714 replied:
> 无法使用 volatile 声明函数，但可以改一下写法让它适配 clang。 #if defined(**clang**) **attribute**((optnone)) #elif defined(**GNUC**) || defined(**GNUG**) **attribute**((optimize("O0"))) #endif

同意，请照此方案改写

Avenger-285714 commented on arch/x86/Kconfig:
```diff
@@ -1515,6 +1515,10 @@ config AMD_MEM_ENCRYPT
 
+config IEE
+	depends on X86_64
+	def_bool y
```
请不要默认使能。

Avenger-285714 commented on arch/x86/kernel/Makefile:
```diff
@@ -162,3 +162,4 @@ ifeq ($(CONFIG_X86_64),y)
 endif
 
 obj-$(CONFIG_HYGON_CSV)			+= csv.o
+obj-$(CONFIG_IEE)				+= haoc/
```
缩进对齐

opsiff commented on arch/arm64/mm/context.c:
```diff
@@ -360,6 +364,18 @@ void cpu_do_switch_mm(phys_addr_t pgd_phys, struct mm_struct *mm)
+	isb();
+#else
 	/* Set ASID in TTBR1 since TCR.A1 is set */
 	ttbr1 &= ~TTBR_ASID_MASK;
```
Clang构建不通过

2025-03-17T08:07:50.7052996Z arch/arm64/mm/context.c:355:16: error: unused variable 'ttbr1' [-Werror,-Wunused-variable]
2025-03-17T08:07:50.7054237Z   355 |         unsigned long ttbr1 = read_sysreg(ttbr1_el1);
2025-03-17T08:07:50.7055014Z       |                       ^~~~~
2025-03-17T08:07:50.7085769Z 1 error generated.
2025-03-17T08:07:50.7184755Z make[4]: *** [scripts/Makefile.build:243：arch/arm64/mm/context.o] 错误 1
2025-03-17T08:07:50.7198380Z make[3]: *** [scripts/Makefile.build:480：arch/arm64/mm] 错误 2
2025-03-17T08:07:50.7202193Z make[3]: *** 正在等待未完成的任务....

amjac27 committed:
HAOC: add support for versions lower than ARMv8.1-a

community inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/IBSGVE

--------------------------------

add support for versions lower than ARMv8.1-a

Fixes: 8d2e6d1bb690 ("HAOC: Add support for AArch64 Isolated Execution Environment(IEE).")
Signed-off-by: Lyu Jinglin <lvjl2022@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

opsiff commented:
评审发现代码有fix未合入 https://gitee.com/openeuler/kernel/pulls/15512

amjac27 committed:
HAOC: Fix some build error when using clang/llvm

We use `#pragma GCC optimize("O0")` because we want to a "memset()" function
for internal use within IEE. However, when the compiler detects operations similar
to memset, it automatically optimizes them into calls to the standard memset
library function. Since IEE calling external functions to access data may cause
security issues, we must prevent this function from being optimized. Now it
supports clang/llvm.

Remove "def_bool = y" for IEE in Kconfig and edit
deepin_x86/arm64_desktop_defconfig to enable IEE by default.

Remove -Wno-unused-variable in Makefile.

Signed-off-by: Lyu Jinglin <lvjl2022@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

amjac27 commented:
1. 已修改代码风格
2. 已取消在 Kconfig 中使能 IEE，改为 savedefconfig 方式
3. 支持 clang/llvm 编译
4. 迁移来自 openeuler 的两个修复补丁（支持1TB以上大内存；识别ARMv8.1-a架构以下的处理器并关闭IEE）
5. 删除 -Wno-unused-variable

Avenger-285714 commented:
/ok-to-test

opsiff commented on arch/arm64/configs/deepin_arm64_desktop_defconfig:
```diff
@@ -155,8 +155,8 @@ CONFIG_ACPI_FFH=y
 CONFIG_KVM=y
+CONFIG_IEE=y
 CONFIG_KPROBES=y
-CONFIG_SHADOW_CALL_STACK=y
```
需要删掉CONFIG_SHADOW_CALL_STACK=y?

Avenger-285714 commented on arch/arm64/include/asm/haoc/iee.h:
```diff
@@ -0,0 +1,72 @@
+#define SET_INVALID(x)	__pgprot(pgprot_val(x) & (~PTE_VALID))
+#define SET_NG(x)	__pgprot(pgprot_val(x) | PTE_NG)
+
+#define PGD_APTABLE_RO	(_AT(pudval_t, 1) << 62)
```
加一下注释说明用法

Avenger-285714 commented on arch/arm64/include/asm/haoc/iee-asm.h:
```diff
@@ -0,0 +1,31 @@
+#define ASID_BIT		(UL(1) << 48)
+
+#ifdef CONFIG_UNMAP_KERNEL_AT_EL0
+#define IEE_ASID			0xfffe
```
这里ASID的使用加注释说明

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/iee-func.c:
```diff
@@ -0,0 +1,318 @@
+			panic("Alloc pgtable error.\n");
+
+		{
+			/* Iterate on the new page table. */
```
调整一下格式

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/iee-func.c:
```diff
@@ -0,0 +1,318 @@
+/* The reverse operation of put_pages_into_iee().
+ * Call this function when you are returning pages back to kernel.
+ */
+void remove_pages_from_iee(unsigned long addr, int order)
```
是否需要加static

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/iee-func.c:
```diff
@@ -0,0 +1,318 @@
+	remove_pages_from_iee(addr, order);
+}
+
+#include <asm/ptrace.h>
```
这一段include为什么没放在顶部

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/iee-mmu.c:
```diff
@@ -0,0 +1,539 @@
+		p4dp = p4d_offset(pgdp, addr);
+		p4d = READ_ONCE(*p4dp);
+		if (!p4d_none(p4d) && !(p4d_val(p4d) & PGD_APTABLE_RO)) {
+			phys += next - addr;
```
这里是否可能导致死循环

Jinglin-lyu replied:
这是在单向依次处理所有 PGD 页表项，每个页表项只可能处理一次，不会导致死循环

Avenger-285714 commented on arch/arm64/kernel/smp.c:
```diff
@@ -461,6 +472,10 @@ void __init smp_prepare_boot_cpu(void)
 	/* Init percpu seeds for random tags after cpus are set up. */
 	kasan_init_sw_tags();
+#ifdef CONFIG_IEE
+	extern void iee_stack_init(void);
```
这一行应该没有用

Avenger-285714 commented on arch/x86/Kconfig:
```diff
@@ -1515,6 +1515,14 @@ config AMD_MEM_ENCRYPT
+		Support for Isolated Execution Environment Framework. Foundation of HAOC.
+		Could isolate kernel critical data and enforce all write access made and
+		verified in IEE APIs.
+	depends on X86_64
```
挪到help上面

Avenger-285714 commented on arch/arm64/kernel/haoc/iee/iee-gate.S:
```diff
@@ -0,0 +1,100 @@
+SYM_FUNC_END(iee_rw_gate)
+
+.global iee_tcr
+iee_tcr:
```
当开启BTF后有以下编译错误：
LD      .tmp_vmlinux.btf
arch/arm64/kernel/haoc/iee/iee-gate.o: in function `iee_protected_rw_gate':
/home/deepin/actions-runner/_work/kernel/kernel/arch/arm64/kernel/haoc/iee/iee-gate.S:55:(.iee.text+0x24): relocation truncated to fit: R_AARCH64_LDST64_ABS_LO12_NC against symbol `iee_tcr' defined in .text section in arch/arm64/kernel/haoc/iee/iee-gate.o
ld: /home/deepin/actions-runner/_work/kernel/kernel/arch/arm64/kernel/haoc/iee/iee-gate.S:55: 警告：one possible cause of this error is that the symbol is being referenced in the indicated code as if it had a larger alignment than was declared where it was defined
/home/deepin/actions-runner/_work/kernel/kernel/arch/arm64/kernel/haoc/iee/iee-gate.S:85:(.iee.text+0x8c): relocation truncated to fit: R_AARCH64_LDST64_ABS_LO12_NC against symbol `kernel_tcr' defined in .text section in arch/arm64/kernel/haoc/iee/iee-gate.o
ld: /home/deepin/actions-runner/_work/kernel/kernel/arch/arm64/kernel/haoc/iee/iee-gate.S:85: 警告：one possible cause of this error is that the symbol is being referenced in the indicated code as if it had a larger alignment than was declared where it was defined
arch/arm64/kernel/haoc/iee/iee-mmu.o: in function `iee_init_tcr':
/home/deepin/actions-runner/_work/kernel/kernel/arch/arm64/kernel/haoc/iee/iee-mmu.c:63:(.init.text+0x2c0): relocation truncated to fit: R_AARCH64_LDST64_ABS_LO12_NC against symbol `kernel_tcr' defined in .text section in arch/arm64/kernel/haoc/iee/iee-gate.o
ld: /home/deepin/actions-runner/_work/kernel/kernel/arch/arm64/kernel/haoc/iee/iee-mmu.c:63: 警告：one possible cause of this error is that the symbol is being referenced in the indicated code as if it had a larger alignment than was declared where it was defined
  BTF     .btf.vmlinux.bin.o
pahole: .tmp_vmlinux.btf: Invalid argument

可能是定义时没有对齐

amjac27 committed:
HAOC: Add some comments and fix build errors

1. add align before "iee_tcr" in iee-gate.S
2. format HAOC Kconfig
3. add some comments for ARM64 iee-rw-gate
4. restore deepin_arm64_desktop_defconfig "CONFIG_SHADOW_CALL_STACK=y"

Signed-off-by: Lyu Jinglin <lvjl2022@zgclab.edu.cn>
Signed-off-by: Liu Zhehui <liuzhh@zgclab.edu.cn>

amjac27 commented:
1. iee-gate.S 中设置对齐，修复编译错误
2. Kconfig 规范修改
3. ARM64 IEE 关键标志位添加注释
4. 恢复 defconfig 中的 CONFIG_SHADOW_CALL_STACK=y
5. 修改提到的代码格式（static，include 等）
